clippy_lints/src/methods/map_unwrap_or_default.rs

```diff
@@ -0,0 +1,73 @@
+use clippy_config::msrvs::{Msrv, OPTION_IS_SOME_AND, RESULT_IS_OK_AND};
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::{expr_sig, is_type_diagnostic_item};
+use rustc_errors::Applicability;
+use rustc_lint::LateContext;
+use rustc_middle::ty::Ty;
+use rustc_span::{sym, Span};
+
+use super::MAP_UNWRAP_OR_DEFAULT;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'_>,
+    expr: &'tcx rustc_hir::Expr<'_>,
+    map_recv: &'tcx rustc_hir::Expr<'_>,
+    map_arg: &'tcx rustc_hir::Expr<'_>,
+    map_span: Span,
+    unwrap_recv: &'tcx rustc_hir::Expr<'_>,
+    msrv: &Msrv,
+) {
+    // Don't lint if:
+
+    // 1. the `expr` is generated by a macro
+    if expr.span.from_expansion() {
+        return;
+    }
+
+    // 2. the caller of `map()` is neither `Option` nor `Result`
+    let is_option = is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(map_recv), sym::Option);
+    let is_result = is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(map_recv), sym::Result);
+    if !is_option && !is_result {
+        return;
+    }
+
+    // 3. the `map_arg` function/closure doesn't return `bool`
+    let map_arg_return_bool = expr_sig(cx, map_arg).is_some_and(|fn_sig| {
+        fn_sig
+            .output()
+            .is_some_and(|ret| ret.no_bound_vars().is_some_and(Ty::is_bool))
+    });
+    if !map_arg_return_bool {
+        return;
+    }
+
+    // 4. msrv doesn't meet either `OPTION_IS_SOME_AND` or `RESULT_IS_OK_AND`
+    if !msrv.meets(OPTION_IS_SOME_AND) || !msrv.meets(RESULT_IS_OK_AND) {
+        return;
+    }
```

Comment:
This doesn't matter since they have the same version, but this is logically the wrong condition. A correct version would first check which function is being suggested and only gate on the function's msrv.

With the previous suggestion combining these it doesn't matter at all.